lib/Sema/CSGen.cpp

```diff
@@ -2051,11 +2051,30 @@ namespace {
     }
 
     FunctionType *inferClosureType(ClosureExpr *closure) {
-      SmallVector<AnyFunctionType::Param, 4> closureParams;
+      // First let's check whether it could be inferred from the body
+      // that some of the untyped parameters have to be `inout`.
+      if (closure->hasSingleExpressionBody()) {
+        auto *body = closure->getSingleExpressionBody();
+        body->forEachChildExpr([&](Expr *expr) -> Expr * {
+          if (auto *assignment = dyn_cast<AssignExpr>(expr)) {
+            if (auto *DRE = dyn_cast<DeclRefExpr>(assignment->getDest())) {
+              if (auto *PD = dyn_cast_or_null<ParamDecl>(DRE->getDecl())) {
```

Comment:
Hmm. This is a little unfortunate. It'll work for the syntactic case where we have an assignment `$0 = foo()`, but won't handle the "obvious-to-users" case like `x.mutatingMethod()`. If we're going to do this, we should at least walk into `TupleExpr` and `ParenExpr` to handle cases like `($0, $1) = foo()`. Was this necessary to fix a bug, or is it speculative?

Reply:
As you mentioned, it's necessary for cases like `_ = { $0 + 1 }`, but I think for member refs there should be a contextual type which supplies the base of the member lookup because without the base we can't lookup what the member is e.g. `_ = { $0.foo = 42 }`, I just couldn't come up with other cases, it probably does make sense for `($0, $1) = foo()` but I completely overlooked that one :)

Reply:
Okay, let's at least do the "destructuring of tuple expressions and paren expressions" part for `($0) = foo()` and `($0, $1) = foo()`.

Reply:
Yeah, absolutely!